mlir/lib/Conversion/ConvertToAIRPass.cpp

```diff
@@ -2318,6 +2318,267 @@ class ScfForallToLaunchConversion : public OpRewritePattern<scf::ForallOp> {
+}
+
+static void extractStridesFromMemrefType(MemRefType memrefTy,
+                                         OpBuilder &builder,
+                                         SmallVector<Value, 4> &strides) {
+  // get the strides and offsets from the memref type
+  int64_t offset;
+  SmallVector<int64_t, 4> layout_strides;
+  auto successStrides = getStridesAndOffset(memrefTy, layout_strides, offset);
+  if (failed(successStrides)) {
+    llvm::outs() << "Failed to get strides\n";
+    return;
+  }
+
+  for (auto s : layout_strides)
+    strides.push_back(
+        builder.create<arith::ConstantIndexOp>(builder.getUnknownLoc(), s));
+}
+
+static void extractSizesFromMemrefType(MemRefType memrefTy, OpBuilder &builder,
+                                       SmallVector<Value, 4> &sizes) {
+  for (auto s : memrefTy.getShape())
+    sizes.push_back(
+        builder.create<arith::ConstantIndexOp>(builder.getUnknownLoc(), s));
+}
+
+static void extractOffsetsFromSubview(memref::SubViewOp subview,
+                                      OpBuilder &builder,
+                                      SmallVector<Value, 4> &offsets) {
+  auto subview_offsets = subview.getOffsets().begin();
+  auto static_offsets = subview.getStaticOffsets();
+  auto loc = subview.getLoc();
+
+  for (auto o : static_offsets) {
+    if (o >= 0)
+      offsets.push_back(builder.create<arith::ConstantIndexOp>(loc, o));
+    else
+      offsets.push_back(*subview_offsets++);
+  }
+}
+
+static LogicalResult canonicalizeAIRDmaOperands(OpBuilder builder,
+                                                SmallVector<Value, 4> &offsets,
+                                                SmallVector<Value, 4> &sizes,
+                                                SmallVector<Value, 4> &strides,
+                                                MemRefType memref) {
+  // Increase vector sizes up to memref size
+  int max_dim_size =
+      std::max(std::max(offsets.size(), sizes.size()), strides.size());
+  if (max_dim_size && offsets.size() < memref.getRank()) {
+    for (unsigned i = offsets.size(); i < memref.getRank(); i++) {
+      offsets.insert(offsets.begin(), builder.create<arith::ConstantIndexOp>(
+                                          builder.getUnknownLoc(), 0));
+    }
+  }
+  if (max_dim_size && sizes.size() < memref.getRank()) {
+    for (unsigned i = sizes.size(); i < memref.getRank(); i++) {
+      sizes.insert(sizes.begin(), builder.create<arith::ConstantIndexOp>(
+                                      builder.getUnknownLoc(), 1));
+    }
+  }
+  int memref_size = 1;
+  for (auto size : memref.getShape())
+    memref_size *= size;
+  if (max_dim_size && strides.size() < memref.getRank()) {
+    for (unsigned i = strides.size(); i < memref.getRank(); i++) {
+      strides.insert(strides.begin(),
+                     builder.create<arith::ConstantIndexOp>(
+                         builder.getUnknownLoc(), memref_size));
```

Comment:
So all dimensions have a stride which is the full size of the tensor? I would have imagined some kind of dimension specific stride (product of all following dimension size). 

Does memref have a static shape (are all elements of memref.getShape() positive)? I'm just trying to connect this to the check for strictly positive values in the static_offsets earlier.

Reply:
At the moment AIR assumes that all strides (one per dimension) are static and that the memref has a static shape. This is to be revised later when an application drives the need for dynamic shapes.

Reply:
Ok. And as for setting all strides to the number of elements in memref, is that because these are 'leading' dimensions?

Reply:
I remember I did that to simply get around a constraint in a downstream AIR pass which complains when the # stride dimensions < memref rank... But I think one bugfix that I did yesterday might have lifted that constraint... Might need to double check.

Anyway, I don't think there's a deep reason for doing that, apart from getting around some engineering constraints...